lib/using_restricted_array.rb

```diff
@@ -6,58 +6,114 @@
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) N represents the number of elements in the array.
+    # The run time will depend on how many elements are in the array
+# Space complexity: O(1) The program will return true or fall no matter the size of the array.
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  x = 0
+  until array[x] == nil
+    return true if array[x] == value_to_find
+    x += 1
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) N represents the number of elements in the array.
+    # The run time will depend on how many elements are in the array
+# Space complexity: O(1). The program will only return the 
+#largest value no matter the size of the array
 def find_largest(array, length)
-  raise NotImplementedError
+  largest = array[0]
+  x = 0
+  until array[x] == nil
+    largest = array[x] if array[x] > largest
+    x += 1
+  end
+  return largest
+  
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1). The program will only return the 
+#smallest value no matter the size of the array
 def find_smallest(array, length)
-  raise NotImplementedError
+  smallest = array[0]
+  x = 0
+  until array[x] == nil
+    smallest = array[x] if array[x] < smallest
+    x += 1
+  end
+  return smallest
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)N represents the number of elements in the array.
+    # The run time will depend on how many elements are in the array
+# Space complexity: O(n) The space needed for the reverse array depends on the size of the original array.
 def reverse(array, length)
-  raise NotImplementedError
+  reverse_array = []
+  x = 0
+  until array[x] == nil
+    reverse_array << array[length-(x+1)]
+    x += 1
+  end
+  return reverse_array
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(log(n))N represents the number of elements in the array.
+    # The run time will depend on how many elements are in the array
+# Space complexity: O(1) The program will return true or fall no matter the size of the array.
 def binary_search(array, length, value_to_find)
-  raise NotImplementedError
-end
+  last = length - 1
+  middle = (last / 2)
+  length/2.times do
+    if array[0] == value_to_find || array[last] == value_to_find
+      return true
+    elsif array[middle] == value_to_find
+      return true
+    elsif array[middle] < value_to_find
+      middle = (middle + last)/2
+    elsif array[middle] > value_to_find
+      middle = (middle / 2)
```

Comment:
This isn't quite right, instead of changing middle directly each time, change the lower index and last index with each iteration and make `middle = (low + last) / 2`